cpp/src/arrow/dataset/scanner_test.cc

```diff
@@ -2758,5 +2789,77 @@ TEST(ScanNode, MinimalGroupedAggEndToEnd) {
   AssertTablesEqual(*expected, *sorted.table(), /*same_chunk_layout=*/false);
 }
 
+TEST(ScanNode, OnlyLoadProjectedFields) {
+  compute::ExecContext exec_context;
+  arrow::dataset::internal::Initialize();
+  ASSERT_OK_AND_ASSIGN(auto plan, compute::ExecPlan::Make());
+
+  auto dummy_schema = schema(
+      {field("key", int64()), field("shared", int64()), field("distinct", int64())});
+
+  // creating a dummy dataset using a dummy table
+  auto table = TableFromJSON(dummy_schema, {R"([
+      [1, 1, 10],
+      [3, 4, 20]
+    ])"});
+
+  auto format = std::make_shared<arrow::dataset::IpcFileFormat>();
+  auto filesystem = std::make_shared<fs::LocalFileSystem>();
+  const std::string file_name = "plan_scan_disk_test.arrow";
+
+  ASSERT_OK_AND_ASSIGN(auto tempdir,
+                       arrow::internal::TemporaryDir::Make("plan-test-tempdir-"));
+  ASSERT_OK_AND_ASSIGN(auto file_path, tempdir->path().Join(file_name));
+  std::string file_path_str = file_path.ToString();
+
+  WriteIpcData(file_path_str, filesystem, table);
+
+  std::vector<fs::FileInfo> files;
+  const std::vector<std::string> f_paths = {file_path_str};
+
+  for (const auto& f_path : f_paths) {
+    ASSERT_OK_AND_ASSIGN(auto f_file, filesystem->GetFileInfo(f_path));
+    files.push_back(std::move(f_file));
+  }
+
+  ASSERT_OK_AND_ASSIGN(auto ds_factory, dataset::FileSystemDatasetFactory::Make(
+                                            filesystem, std::move(files), format, {}));
+  ASSERT_OK_AND_ASSIGN(auto dataset, ds_factory->Finish(dummy_schema));
+
+  auto scan_options = std::make_shared<dataset::ScanOptions>();
+  compute::Expression extract_expr = compute::field_ref("shared");
+  // don't use a function.
+  scan_options->projection =
+      call("make_struct", {extract_expr}, compute::MakeStructOptions{{"shared"}});
+
+  arrow::AsyncGenerator<std::optional<compute::ExecBatch>> sink_gen;
+
+  auto declarations = compute::Declaration::Sequence(
+      {compute::Declaration({"scan", dataset::ScanNodeOptions{dataset, scan_options}}),
+       compute::Declaration({"sink", compute::SinkNodeOptions{&sink_gen}})});
+
+  ASSERT_OK_AND_ASSIGN(auto decl, declarations.AddToPlan(plan.get()));
+
+  ASSERT_OK(decl->Validate());
+
+  auto output_schema =
+      schema({field("a", int64()), field("b", int64()), field("c", int64())});
+
+  auto expected = TableFromJSON(dummy_schema, {R"([
+      [null, 1, null],
+      [null, 4, null]
+  ])"});
+
+  std::shared_ptr<arrow::RecordBatchReader> sink_reader = compute::MakeGeneratorReader(
+      output_schema, std::move(sink_gen), exec_context.memory_pool());
+
+  ASSERT_OK(plan->Validate());
+  ASSERT_OK(plan->StartProducing());
+  ASSERT_OK_AND_ASSIGN(auto actual,
+                       arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+  AssertTablesEqual(*expected, *actual, /*same_chunk_layout=*/false);
```

Comment:
```suggestion

  auto declarations = compute::Declaration::Sequence(
      {compute::Declaration({"scan", dataset::ScanNodeOptions{dataset, scan_options}})});

  ASSERT_OK_AND_ASSIGN(auto actual, compute::DeclarationToTable(declarations));
  // Scan node always emits augmented fields so we drop those
  ASSERT_OK_AND_ASSIGN(auto actualMinusAgumented, actual->SelectColumns({0, 1, 2}));

  auto expected = TableFromJSON(dummy_schema, {R"([
      [null, 1, null],
      [null, 4, null]
  ])"});

  AssertTablesEqual(*expected, *actualMinusAgumented, /*same_chunk_layout=*/false);
```

We have `DeclarationToTable` now which is more succinct.  Also, the scan node was attaching augmented columns but you were not seeing them because the sink reader's schema didn't have all the columns.